src/config.py

```diff
@@ -0,0 +1,274 @@
+
+        Returns:
+            True if the repository was already synced or synced successfully, False otherwise
+        """
+        repo_cloned = self.url == self._get_cloned_repo_url()
+        repo_ref_matched = self.ref == self._get_cloned_repo_ref()
+        if repo_cloned and repo_ref_matched:
+            logger.info("Setting repository already synced")
+            return True
+
+        if not self.url or not self.host:
+            return False
+
+        self._add_ssh_key()
+        self._add_known_hosts()
+
+        if self.working_dir.exists():
+            shutil.rmtree(self.working_dir)
+
+        self._git_clone()
+
+        sub_path = self.sub_path or ""
+        rules_dir = self.working_dir / sub_path / RSYNC_CUSTOM_RULES_KEY
+        configs_dir = self.working_dir / sub_path / RSYNC_CUSTOM_CONFIGS_KEY
+        if rules_dir.exists():
+            self._rsync(f"{rules_dir}/", f"{falco_layout.rules_dir}/")
+        if configs_dir.exists():
+            self._rsync(f"{configs_dir}/", f"{falco_layout.configs_dir}/")
+        return True
+
+    def _rsync(self, source: str, destination: str) -> None:
+        """Rsync files from source to destination."""
+        rsync_cmd = [RSYNC, "-av", "--delete", source, destination]
+        try:
+            subprocess.run(rsync_cmd, check=True)
+        except subprocess.CalledProcessError as e:
+            logging.exception("Rsync failed from %s to %s", source, destination)
+            raise RsyncError(f"Rsync failed: {e.stderr}") from e
+
+    def _add_ssh_key(self) -> None:
+        """Add the Ssh private key to the host."""
+        Ssh_DIR.mkdir(mode=0o700, exist_ok=True)
+        try:
+            with SSH_KEY_FILE.open("w", encoding="utf-8") as key_file:
+                key_file.write(self.ssh_private_key)
+            SSH_KEY_FILE.chmod(0o600)
+        except OSError as e:
+            logging.exception("Error writing Ssh key to %s", SSH_KEY_FILE)
+            raise SshKeyWriteError(f"Error writing Ssh key to {SSH_KEY_FILE}") from e
+
+    def _add_known_hosts(self) -> None:
+        """Scan and add the Ssh host key to known_hosts.
+
+        Raises:
+            SshKeyScanError: If ssh-keyscan fails
+        """
+        add_known_hosts_cmd = [SSH_KEYSCAN, "-t", "rsa", self.host]
+        try:
+            out = subprocess.check_output(add_known_hosts_cmd).decode()
+            with KNOWN_HOSTS_FILE.open("w", encoding="utf-8") as known_hosts_file:
+                known_hosts_file.write(out)
+        except subprocess.CalledProcessError as e:
+            logging.exception("Ssh keyscan failed for host %s", self.host)
+            raise SshKeyScanError(f"Ssh keyscan failed: {e.stderr}") from e
+        except OSError as e:
+            logging.exception("Error writing to known_hosts at %s", KNOWN_HOSTS_FILE)
+            raise SshKeyScanError(f"Error writing to known_hosts at {KNOWN_HOSTS_FILE}") from e
+
+    def _git_clone(self) -> None:
```

Comment:
Can you move the functionality to another class? Our convention is to use a class to strictly validate the configuration and relation data